tests/nix/nat-testing.nix

```diff
@@ -0,0 +1,128 @@
+# here is the testing base file: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/testing-python.nix
+with import <nixpkgs> {};
+let
+  pk = (callPackage ../../nix/default.nix {}).package;
+in
+  import <nixpkgs/nixos/tests/make-test-python.nix> {
+    nodes =
+      {
+        privateNode1 =
+          { nodes, pkgs, ... }:
+          {
+            virtualisation.vlans = [ 1 ];
+            environment.variables = {
+              KEYNODE_PATH = "$HOME/.polykey";
+            };
+            environment.systemPackages = [ pk pkgs.tcpdump ];
+            networking.firewall.enable = false;
+            networking.defaultGateway = (pkgs.lib.head nodes.router1.config.networking.interfaces.eth1.ipv4.addresses).address;
+          };
+        privateNode2 =
+          { nodes, pkgs, ... }:
+          {
+            virtualisation.vlans = [ 2 ];
+            environment.variables = {
+              KEYNODE_PATH = "$HOME/.polykey";
+            };
+            environment.systemPackages = [ pk pkgs.tcpdump ];
+            networking.firewall.enable = false;
+            networking.defaultGateway = (pkgs.lib.head nodes.router2.config.networking.interfaces.eth1.ipv4.addresses).address;
+          };
+        router1 =
+          { pkgs, ... }:
+          {
+            virtualisation.vlans = [ 1 3 ];
+            environment.systemPackages = [ pkgs.tcpdump ];
+            networking.firewall.enable = false;
+            networking.nat.externalInterface = "eth2";
+            networking.nat.internalIPs = [ "192.168.1.0/24" ];
+            networking.nat.enable = true;
+          };
+        router2 =
+          { ... }:
+          {
+            virtualisation.vlans = [ 2 3 ];
+            environment.systemPackages = [ pkgs.tcpdump ];
+            networking.firewall.enable = false;
+            networking.nat.externalInterface = "eth2";
+            networking.nat.internalIPs = [ "192.168.2.0/24" ];
+            networking.nat.enable = true;
+          };
+        publicNode =
+          { config, pkgs, ... }:
+          {
+            virtualisation.vlans = [ 3 ];
+            environment.variables = {
+              KEYNODE_PATH = "$HOME/.polykey";
+            };
+            environment.systemPackages = [ pk pkgs.tcpdump ];
+            networking.firewall.enable = false;
+          };
+      };
+    testScript =''
```

Comment:
This script can be put outside of the Nix file and you can use `readFile`. The script is basically python right? Also the whole thing requires a diagram in a README.md.

This would probably be the test hierarchy;

```
tests/bin/just run the CLI executables (basic workflow behaviour)
tests/unit/<module> <- really just unit testing with mocking and DI (if you can DI, no need for mocking)
tests/network/
```

Where `tests/network/` involve network simulation.